source/extensions/filters/http/credential_injector/credential_injector_filter.cc

```diff
@@ -0,0 +1,100 @@
+#include "source/extensions/filters/http/credential_injector/credential_injector_filter.h"
+
+#include "source/common/http/headers.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace CredentialInjector {
+
+FilterConfig::FilterConfig(CredentialInjectorSharedPtr credential_injector, bool overwrite,
+                           bool allow_request_without_credential, const std::string& stats_prefix,
+                           Stats::Scope& scope)
+    : injector_(credential_injector), overwrite_(overwrite),
+      allow_request_without_credential_(allow_request_without_credential),
+      stats_(generateStats(stats_prefix + "credential_injector.", scope)) {}
+
+// Inject configured credential to the HTTP request header.
+// return should continue processing the request or not
+bool FilterConfig::injectCredential(Http::RequestHeaderMap& headers) {
+  absl::Status status = injector_->inject(headers, overwrite_);
+
+  // credential already exists in the header and overwrite is false
+  if (absl::IsAlreadyExists(status)) {
+    ASSERT(!overwrite_); // overwrite should be false if AlreadyExists is returned
+    ENVOY_LOG(warn, "Credential already exists in the header");
+    stats_.already_exists_.inc();
+    // continue processing the request if credential already exists in the header
+    return true;
+  }
+
+  // failed to inject the credential
+  if (!status.ok()) {
+    ENVOY_LOG(warn, "Failed to inject credential: {}", status.message());
```

Comment:
I think we need to log at `debug` here and above since this code gets executed per-request.